scvi/external/sysvi/_base_components.py

```diff
@@ -0,0 +1,347 @@
+    Dropout,
+    LayerNorm,
+    Linear,
+    Module,
+    Parameter,
+    ReLU,
+    Sequential,
+)
+
+
+class Embedding(Module):
+    """Module for obtaining embedding of categorical covariates
+
+    Parameters
+    ----------
+    size
+        N categories
+    cov_embed_dims
+        Dimensions of embedding
+    normalize
+        Apply layer normalization
+    """
+
+    def __init__(self, size, cov_embed_dims: int = 10, normalize: bool = True):
+        super().__init__()
+
+        self.normalize = normalize
+
+        self.embedding = torch.nn.Embedding(size, cov_embed_dims)
+
+        if self.normalize:
+            # TODO this could probably be implemented more efficiently as embed gives same result for every sample in
+            #  a give class. However, if we have many balanced classes there wont be many repetitions within minibatch
+            self.layer_norm = torch.nn.LayerNorm(
+                cov_embed_dims, elementwise_affine=False
+            )
+
+    def forward(self, x):
+        x = self.embedding(x)
+        if self.normalize:
+            x = self.layer_norm(x)
+
+        return x
+
+
+class EncoderDecoder(Module):
+    """Module that can be used as probabilistic encoder or decoder
+
+    Based on inputs and optional covariates predicts output mean and var
+
+    Parameters
+    ----------
+    n_input
+    n_output
+    n_cov
+    n_hidden
+    n_layers
+    var_eps
+        See :class:`~scvi.external.sysvi.nn.VarEncoder`
+    var_mode
+        See :class:`~scvi.external.sysvi.nn.VarEncoder`
+    sample
+        Return samples from predicted distribution
+    kwargs
+        Passed to :class:`~scvi.external.sysvi.nn.Layers`
+    """
+
+    def __init__(
+        self,
+        n_input: int,
+        n_output: int,
+        n_cov: int,
+        n_hidden: int = 256,
+        n_layers: int = 3,
+        var_eps: float = 1e-4,
+        var_mode: str = "feature",
+        sample: bool = False,
+        **kwargs,
+    ):
+        super().__init__()
+        self.sample = sample
+
+        self.var_eps = var_eps
+
+        self.decoder_y = Layers(
+            n_in=n_input,
+            n_cov=n_cov,
+            n_out=n_hidden,
+            n_hidden=n_hidden,
+            n_layers=n_layers,
+            **kwargs,
+        )
+
+        self.mean_encoder = Linear(n_hidden, n_output)
+        self.var_encoder = VarEncoder(n_hidden, n_output, mode=var_mode, eps=var_eps)
+
+    def forward(self, x, cov: torch.Tensor | None = None):
+        y = self.decoder_y(x=x, cov=cov)
+        # TODO better handling of inappropriate edge-case values than nan_to_num or at least warn
+        y_m = torch.nan_to_num(self.mean_encoder(y))
+        y_v = self.var_encoder(y, x_m=y_m)
+
+        outputs = {"y_m": y_m, "y_v": y_v}
+
+        # Sample from latent distribution
+        if self.sample:
+            y = Normal(y_m, y_v.sqrt()).rsample()
+            outputs["y"] = y
+
+        return outputs
+
+
+class Layers(Module):
```

Comment:
Since this class is similar to the existing implementation of `FCLayers`, could you refactor `Layers` so that it subclasses `FCLayers` and overrides methods as needed? This will make it clearer what's different about this implementation as well as remove duplicate code in `inject_into_layer` and `set_online_update_hooks`, as it looks like these methods are identical to the original

Reply:
Inheriting this will make it also easier for us to add new functionality to sysVI. Are there issues with the FClayers class?